feat: Add macOS ARM64 platform support

johann-taberlet opened the pull request:
## Summary
Adds full macOS ARM64 (Apple Silicon) support to the GDExtension, including build system changes and critical bug fixes discovered during testing.

## Changes

### Build System
- Updated SCons/CMake to support macOS ARM64 with vcpkg
- Pre-built litehtml/gumbo libraries for macOS (linked with `-force_load` for proper static initialization)
- Added abseil linking for macOS builds
- Added macOS ARM64 entries to `Godot3DTiles.gdextension`

### HTTP Client
- New `GodotHttpClient` using Godot's native HTTPClient (avoids curl SSL/SNI issues on macOS)
- Handles `data:` URIs for Cesium credit system logos (base64-encoded inline images)
- Used automatically on macOS for all HTTP requests (NetworkAssetAccessor, DocumentContainer, TokenTroubleShooting)

### Bug Fixes (discovered during macOS testing, benefits all platforms)
- **Fixed dangling pointer crash**: Model pointer was captured before `tileLoadResult` was moved in lambda, causing all tiles to fail with "Model has no nodes"
- **Fixed null dereference in prepareInMainThread**: Added check when tile loading fails
- **Fixed null dereference in despawn_tile**: Added check when `getRenderResources()` returns null
- **Fixed vector exception**: Added bounds check for optional glTF scale array
- **Fixed promise rejection crashes**: Use `resolve()` instead of `reject()` with `LIBASYNC_NO_EXCEPTIONS`

## Test Plan
- [x] Build on macOS ARM64: `scons arch=arm64 compileTarget=extension target=template_release`
- [x] Install plugin in Godot 4.5 project
- [x] Connect to Cesium Ion via plugin UI
- [x] Verify Cesium World Terrain renders correctly
- [x] Verify Google Photorealistic 3D Tiles renders correctly
- [x] Verify OSM Buildings renders correctly (collision generation has pre-existing issues)
- [x] Verify Cesium credit images load properly

## Notes
- The bug fixes in `GodotPrepareRenderResources.cpp` and `CesiumGDTileset.cpp` are not macOS-specific and improve stability on all platforms
- OSM Buildings collision generation fails with complex meshes - this is a pre-existing issue unrelated to macOS support

johann-taberlet committed:
feat: Add macOS ARM64 build system support

- Add macOS platform detection using sys.platform == 'darwin'
- Add arm64-osx vcpkg triplet for Apple Silicon
- Add macOS-specific compile flags (-std=c++20, -fexceptions, -fPIC)
- Add build_litehtml_macos() to compile litehtml from source
- Fix macOS shared library output path in SConstruct
- Add macOS library linking with frameworks (CoreFoundation, Security)
- Add macOS litehtml library path

johann-taberlet committed:
feat: Add GodotHttpClient using native Godot HTTPClient

Replace curl with Godot's built-in HTTPClient for macOS to avoid
DNS resolution and SSL/SNI issues that occur with libcurl on macOS.

- Uses Godot's HTTPClient with proper TLS configuration
- Handles data: URIs for base64-encoded inline images (credit logos)
- Thread pool for concurrent requests (same pattern as CurlHttpClient)
- Proper URL parsing with scheme, host, port, and path extraction
- Timeout handling for connection and request phases

johann-taberlet committed:
refactor: Use GodotHttpClient on macOS in NetworkAssetAccessor

Conditionally compile NetworkAssetAccessor to use:
- GodotHttpClient on macOS (native Godot HTTPClient)
- CurlHttpClient on Windows/Linux (libcurl)

This hybrid approach keeps the existing curl-based implementation
working on platforms where it's stable while using Godot's native
HTTP on macOS where curl has issues.

johann-taberlet committed:
fix: Replace promise reject() with resolve() to prevent crashes

With LIBASYNC_NO_EXCEPTIONS defined, calling promise.reject() triggers
std::terminate() which crashes the application. This is because the
async++ library uses exceptions internally for rejection handling.

Changes:
- NetworkAssetAccessor: Always resolve with HTTP status codes (400, etc.)
- LocalAssetAccesor: Return 404 for file not found, 501 for not implemented
- GodotPrepareRenderResources: Resolve with nullptr on mesh generation error

Cesium Native properly handles error responses based on status codes,
so this approach maintains correct error handling while avoiding crashes.

johann-taberlet committed:
feat: Add macOS ARM64 entries to GDExtension configuration

Add library paths for macOS ARM64 (Apple Silicon) in both debug
and release configurations.

johann-taberlet committed:
fix: Fix macOS build system issues

- Add SHLIBSUFFIX to macOS library output path
- Add is_macos() checks to prevent Linux-specific linker flags
  (os.name == 'posix' on both Linux and macOS)

johann-taberlet committed:
fix: Fix type compatibility issues for macOS ARM64

- Change size_t to uint64_t in free_tile() signature for godot-cpp
  compatibility (size_t maps to unsigned long on macOS which isn't
  supported by PtrToArg)
- Fix designated initializer in CesiumPropertyInfo which fails on
  non-aggregate types with Apple Clang

johann-taberlet committed:
fix: macOS ARM64 linking with abseil and force_load for litehtml

- Add abseil libraries required by s2 geometry on macOS
- Use -force_load for litehtml/gumbo to resolve internal archive
  dependencies (macOS linker doesn't handle this automatically)
- Include pre-built macOS ARM64 litehtml static libraries
- Separate litehtml path handling for macOS vs other platforms

johann-taberlet committed:
fix: Resolve tile loading crashes from dangling pointer and null derefs

- Fix dangling pointer in prepareInLoadThread: re-extract model pointer
  from tileLoadResult after it's moved into the lambda capture
- Add null check in prepareInMainThread for failed load thread results
- Add null check in despawn_tile before calling is_inside_tree()
- Add bounds check for optional glTF scale array (defaults to 1,1,1)

These fixes prevent crashes when tile loading fails gracefully.

johann-taberlet committed:
feat: Use GodotHttpClient for all HTTP on macOS

- DocumentContainer: Use GodotHttpClient for credit image loading
- TokenTroubleShooting: Use GodotHttpClient for token validation
- GodotHttpClient: Add send_get_same_thread method for synchronous requests

libcurl has DNS resolution and SSL/SNI issues on macOS. Using Godot's
native HTTPClient provides better compatibility with macOS networking.

johann-taberlet committed:
fix: Correct macOS debug library path in gdextension config

Use template_debug instead of template_release for macos.debug.arm64
to match the actual built library filename.

Kyn21kx commented:
Excellent work, going into review now

Kyn21kx commented on CesiumBuildUtils.py:
```diff
@@ -24,16 +24,25 @@
 
 
+def is_macos():
+    return sys.platform == PLATFORM_MACOS
```
This function is redundant, the check of `PLATFORM_MACOS` or `OS_MACOS` should be enough

Kyn21kx commented on CesiumBuildUtils.py:
```diff
@@ -24,16 +24,25 @@
 def get_compile_flags():
     if os.name == OS_WIN:
         return ["/std:c++20", "/Zc:__cplusplus", "/utf-8", "/bigobj"]
+    elif is_macos():
```
We can probably switch the entire platform checks from `os.name` to `sys.platform`, this is something I can do once the PR is merged, or, if you want to take on that task it's fine by me

Kyn21kx commented on CesiumBuildUtils.py:
```diff
@@ -148,6 +157,70 @@ def clone_lite_html_if_needed():
     pass
 
 
+def build_litehtml_macos():
```
Change this function name to properly reflect the architecture for which it's going to be built (arm64 in this case), or add a parameter to configure the architecture dynamically

Kyn21kx commented on CesiumBuildUtils.py:
```diff
@@ -148,6 +157,70 @@ def clone_lite_html_if_needed():
+        return
+
+    # Copy output libraries
+    import shutil
```
Prefer imports at the top level of the function or the global scope

Kyn21kx commented on cesium_godot/Utils/GodotHttpClient.h:
```diff
@@ -0,0 +1,262 @@
+		if (portPos != std::string::npos) {
+			parts.host = hostPart.substr(0, portPos);
+			try {
+				parts.port = std::stoi(hostPart.substr(portPos + 1));
```
Exceptions are strongly discouraged in this codebase, prefer `std::from_chars` instead of `std::stoi`

Kyn21kx commented on cesium_godot/Utils/GodotHttpClient.h:
```diff
@@ -0,0 +1,262 @@
+			}
+		} else {
+			parts.host = hostPart;
+			parts.port = parts.use_tls ? 443 : 80;
```
Define these ports as named `constexpr`

Kyn21kx commented on cesium_godot/Utils/GodotHttpClient.h:
```diff
@@ -0,0 +1,262 @@
+				std::string base64Data = urlStr.substr(base64Pos + 8);
+				String godotBase64 = String(base64Data.c_str());
+				buffer = Marshalls::get_singleton()->base64_to_raw(godotBase64);
+				*outStatus = 200;
```
Prefer using Godot's underlying `HTTPClient::ResponseCode` enum instead of magic numbers

Kyn21kx commented on cesium_godot/Utils/GodotHttpClient.h:
```diff
@@ -0,0 +1,262 @@
+				return buffer;
+			}
+			ERR_PRINT(String("Invalid data URI format: ") + url);
+			*outStatus = 400;
```
Same thing, use `HTTPClient::ResponseCode`

Kyn21kx commented on cesium_godot/Utils/GodotHttpClient.h:
```diff
@@ -0,0 +1,262 @@
+		UrlParts parts = parse_url(urlStr);
+		if (!parts.valid) {
+			ERR_PRINT(String("Invalid URL format: ") + url);
+			*outStatus = 400;
```
Same thing `HTTPClient::ResponseCode`

Kyn21kx commented on cesium_godot/Utils/GodotHttpClient.h:
```diff
@@ -0,0 +1,262 @@
+		}
+
+		// Poll until connected (max ~10 seconds)
+		constexpr int MAX_CONNECT_POLLS = 1000;
```
Avoid unspecified integers, prefer typedefs in `cstdint` such as `int32_t`

Kyn21kx commented on cesium_godot/Utils/GodotHttpClient.h:
```diff
@@ -0,0 +1,262 @@
+		parts.scheme = url.substr(0, schemeEnd);
+		parts.use_tls = (parts.scheme == "https");
+
+		size_t hostStart = schemeEnd + 3;
```
Avoid magic numbers for offsets

Kyn21kx reviewed (changes requested):
Great changes, just a few conventions that need to be followed to ensure the maintainability of the project

JadeWhite commented:
Hey, are you still working on this? I’d love to help push it over the line if not.

Kyn21kx commented:
> Hey, are you still working on this? I’d love to help push it over the line if not.

Hi there, I think work on this branch has been stalled for a bit, if you'd like to contribute you're welcome to do so and I'll follow up as closely as possible :)

johann-taberlet committed:
fix: Address PR review feedback on code conventions

- Remove redundant is_macos() helper, use sys.platform == PLATFORM_MACOS directly
- Rename build_litehtml_macos() to build_litehtml(arch) for dynamic architecture
- Move shutil import to module top level
- Replace std::stoi with std::from_chars (no exceptions)
- Define named constexpr for ports and URL scheme offset
- Use HTTPClient::ResponseCode enum instead of magic numbers
- Use int32_t instead of unspecified int

johann-taberlet committed:
Merge upstream/master into feat/macos-support

Resolve conflict in NetworkAssetAccessor.cpp: keep mutable lambda
(LIBASYNC_NO_EXCEPTIONS fix) and upstream's file:// URL support +
improved error logging.

johann-taberlet commented:
Hey, sorry for the long delay on this! I finally got around to addressing all the review feedback:

- Removed the redundant `is_macos()` helper, using `sys.platform == PLATFORM_MACOS` directly
- Renamed `build_litehtml_macos()` to `build_litehtml(arch)` with a configurable architecture parameter
- Moved the `shutil` import to module top level
- Replaced `std::stoi` with `std::from_chars` (no exceptions)
- Added named `constexpr` for ports and URL scheme offset
- Switched to `HTTPClient::ResponseCode` enum instead of magic numbers
- Used `int32_t` instead of unspecified `int`

Also merged latest master to resolve the conflict.

Let me know if there's anything else that needs fixing!

DarshanDoesStuff commented:
I just spent the last 2 days trying to get this working on my system just for me to check back for the first time in forever today and find out it was reviewed 4 hours ago. 🤣
